modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java

```diff
@@ -275,18 +274,16 @@ public boolean compatibilityMode() {
 
                 if (fut != null) {
                     if (asyncWaitForTransition) {
-                         return new IgniteFutureImpl<>(fut.chain(new C1<IgniteInternalFuture<Void>, Boolean>() {
-                            @Override public Boolean apply(IgniteInternalFuture<Void> fut) {
-                                Boolean res = globalState.transitionResult();
+                        return new IgniteFutureImpl<>(fut.chain((C1<IgniteInternalFuture<Void>, ClusterState>)f -> {
+                            ClusterState res = globalState.transitionResult();
 
-                                assert res != null;
+                            assert res != null;
 
-                                return res;
-                            }
+                            return res;
                         }));
                     }
                     else
-                        return new IgniteFinishedFutureImpl<>(globalState.previouslyActive());
+                        return new IgniteFinishedFutureImpl<>(lesserOf(globalState.lastState(), globalState.state()));
```

Comment:
Value returned in future doesn't match description from the javadoc: 
  ```
 /**
     * @param waitForTransition Wait end of transition or not.
     * @return Current cluster state to be used on public API.
     */
    IgniteFuture<ClusterState> publicApiStateAsync(boolean waitForTransition);
```

Reply:
Please take a look at http://apache-ignite-developers.2346864.n4.nabble.com/DISCUSSION-Single-point-in-API-for-changing-cluster-state-tt43665.html#a44196 
We should return state with lesser functionality for avoid unexpected exceptions on user side.

Reply:
Got it, it's logic for waitForTransition = false